docs/hello.rst

```diff
@@ -140,84 +141,108 @@ For those who came from the world of CUDA, ``ti.func`` corresponds to ``__device
-**Struct-for loops** have a cleaner syntax, and are particularly useful when iterating over tensor elements.
-In the fractal code above, ``for i, j in pixels`` loops over all the pixel coordinates, i.e. ``(0, 0), (0, 1), (0, 2), ... , (0, 319), (1, 0), ..., (639, 319)``.
+    .. code-block:: python
+
+        @ti.kernel
+        def foo():
+            for i in range(10): # Parallelized :-)
+                ...
+
+        @ti.kernel
+        def bar(k: ti.i32):
+            if k > 42:
+                for i in range(10): # Serial :-(
+                    ...
+
+**Struct-for loops** have are particularly useful when iterating over (sparse) tensor elements.
+In the code above, ``for i, j in pixels`` loops over all the pixel coordinates, i.e. ``(0, 0), (0, 1), (0, 2), ... , (0, 319), (1, 0), ..., (639, 319)``.
 
 .. note::
 
     Struct-for is the key to :ref:`sparse` in Taichi, as it will only loop over active elements in a sparse tensor. In dense tensors, all elements are active.
 
-.. note::
-    Struct-for's must be at the outer-most scope of kernels.
+.. warning::
+
+    Struct-for's must live at the outer-most scope of kernels.
 
-.. note::
     It is the loop **at the outermost scope** that gets parallelized, not the outermost loop.
 
     .. code-block:: python
 
-      # Good kernel
-      @ti.func
-      def foo():
-        for i in x:
-          ...
-
-      # Bad kernel
-      @ti.func
-      def bar(k: ti.i32):
-        # The outermost scope is a `if` statement, not the struct-for loop!
-        if k > 42:
-          for i in x:
-            ...
+        @ti.kernel
+        def foo():
+            for i in x:
+                ...
+
+        @ti.kernel
+        def bar(k: ti.i32):
+            # The outermost scope is a `if` statement
+            if k > 42:
+                for i in x: # Not allowed. Struct-fors must live in the outermost scope.
+                    ...
 
-.. note::
-    ``break`` is not supported in **outermost (parallelized)** loops:
+
+
+
+.. warning::
+
+    ``break`` **is not supported in parallel loops:**
 
     .. code-block:: python
 
       @ti.kernel
       def foo():
         for i in x:
             ...
-            break # ERROR! You cannot break a parallelized loop!
+            break # Error!
+
+        for i in range(10):
+            ...
+            break # Error!
 
       @ti.kernel
       def foo():
         for i in x:
-            for j in y:
+            for j in range(10):
                 ...
-                break # OK
+                break # OK!
+            break # Error!
 
 
 Interacting with Python
 ------------------------
 
-Everything outside Taichi-scope (``ti.func`` and ``ti.kernel``) is simply Python. You can use your favorite Python packages (e.g. ``numpy``, ``pytorch``, ``matplotlib``) with Taichi.
+Everything outside Taichi-scopes (``ti.func`` and ``ti.kernel``) is simply Python. You can use your favorite Python packages (e.g. ``numpy``, ``pytorch``, ``matplotlib``) with Taichi.
```

Comment:
```suggestion
Everything outside Taichi-scope (``ti.func`` and ``ti.kernel``) is simply Python. You can use your favorite Python packages (e.g. ``numpy``, ``pytorch``, ``matplotlib``) with Taichi.
```

Reply:
I'd actually stick to plural forms here since we may have multiple `ti.kernel` and each is a Taichi-scope. WDYT?